### ISSUE TITLE
audience value is different after Upgrade from ID4 to Duende Identity Server

### ISSUE THREAD

**Issue by IAMHK90** [DuendeSoftware/Support#825]
Duende IdentityServer : 6.3.3
.net6

With ID4, the token included a list of all the resources in the audience array. 

<img width="320" alt="image" src="https://github.com/DuendeSoftware/Support/assets/31859932/fb3ae2aa-b2b6-4e82-876c-482ef30bbd46">

With Duende Identity Server 6.3.3, I'm not seeing the audience list.

<img width="344" alt="image" src="https://github.com/DuendeSoftware/Support/assets/31859932/78c400a8-6b11-4e9c-8465-72683774c123">

Am I missing anything?

**Comment by josephdecock** [comment]
There's quite a bit of history around the audience claim, because various specifications have had conflicting guidance over the years. Dominick blogged about the issues [here](https://leastprivilege.com/2020/06/18/resource-access-in-identityserver4-v4-and-going-forward/).

It looks like you need to add [ApiResource](https://docs.duendesoftware.com/identityserver/v6/fundamentals/resources/api_resources/) configuration to your IdentityServer. You can describe the scopes that are supported by the different resources in your configuration as well. Then when you request those scopes, the associated resources should appear as audiences in your token.

**Comment by IAMHK90** [comment]
AddConfigurationStore must get the config data from DB including  apiresources right? If yes, then we already added that.

**Comment by josephdecock** [comment]
Yes, AddConfigurateionStore will get config data from the database for your ApiResources. Are you requesting scopes that are supported by the ApiResources that you're expecting? Each ApiResource includes a collection of scopes, so you need to make sure that your resources include the scopes that you're requesting. Also, ApiResources have a flag "RequireResourceIndicator". That should be off on your resources, unless you are using resource indicators (probably not, since you're migrating from an older version that didn't do that).

**Comment by IAMHK90** [comment]
Scopes are included for each APiResource. I did set RequireResourceIndicator to false as well. I still see the audience value in string instead of an array which gives us the list of all resources/scopes.

**Comment by IAMHK90** [comment]
@josephdecock,  is there something further you would suggest on this? any help would be much appreciated!

**Comment by josephdecock** [comment]
We do have a sample [here](https://github.com/DuendeSoftware/Samples/tree/main/IdentityServer/v6/ScopesAndResources) that shows how the scope and resource parameters interact to change the audiences and scopes of the tokens that are produced. You could try experimenting with that and comparing the configuration there with your configuration to see if anything sticks out.

**Comment by IAMHK90** [comment]
I found it! Map the resources and scopes in [ApiResourceScopes] table. I did it only in [ApiScopes] table.  [ApiScopes] table already has the mapping with the resources (ApiResourceId).  One thing that is not clear to me is why should we map resources and scopes in two tables (ApiResourceScopes, and ApiScopes)?

**Comment by josephdecock** [comment]
My guess is that your ApiScope's ApiResourceId column is left over from before your upgrade.

In IdentityServer4 v3 there was a parent-child relationship between the ApiResources and the ApiScopes. Then in IdentityServer4 v4 the ApiScopes was promoted to be its own top-level configuration. This meant that the child collection under the ApiResources was renamed to ApiResourcesScopes and it contained a reference to the new top-level ApiScopes.

**Comment by IAMHK90** [comment]
Can we say that it is safe to remove ApiResourceId column from ApiScope's table?

**Comment by josephdecock** [comment]
Yes, it is no longer part of the in the ApiScope model or entity.

**Comment by kapilkhub** [comment]
Hi, I have a similar question. 

I too only getting "/resources" as aud in token. 

I have already mapped ApiScopes and ApiResources in ApiResourceScopesTable.

![Image](https://github.com/user-attachments/assets/4c5af029-2128-4601-8006-fb16292f94e9)


I also added openid, and profile in identityresources table.  Really appreciate your help   

not sure what I am missing.

**Comment by josephdecock** [comment]
@kapilkhub would you please create a new support issue with the details you've shared, as well as adding more info about the behavior you're experiencing vs what you expected?
